sherpa/csrc/byte_util.cc

```diff
@@ -0,0 +1,212 @@
+  }
+  return oss.str();
+}
+
+void ByteUtil::UTF8StringToBytes(const std::string &str,
+                                 std::vector<uint8_t> *bytes) const {
+  const char *data = str.data();
+  const size_t length = str.size();
+  for (size_t i = 0; i < length; /* no update */) {
+    int32_t c = data[i++] & 0xff;
+    if ((c & 0x80) == 0) {
+      if (c > max_token_ || token2byte_[c] == -1) {
+        SHERPA_LOG(WARNING) << "Skip OOV token, code point : " << c
+                            << " utf8 char : " << CodePointToUTF8String(c);
+        continue;
+      }
+      bytes->push_back(token2byte_[c]);
+    } else {
+      if ((c & 0xc0) == 0x80) {
+        SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+      }
+      int32_t count =
+          (c >= 0xc0) + (c >= 0xe0) + (c >= 0xf0) + (c >= 0xf8) + (c >= 0xfc);
+      int32_t code = c & ((1 << (6 - count)) - 1);
+      while (count != 0) {
+        if (i == length) {
+          SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+        }
+        char cb = data[i++];
+        if ((cb & 0xc0) != 0x80) {
+          SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+        }
+        code = (code << 6) | (cb & 0x3f);
+        count--;
+      }
+      if (code < 0) {
+        // This should not be able to happen.
+        SHERPA_LOG(FATAL) << "Invalid utf8 string.";
+      }
+      if (code > max_token_ || token2byte_[code] == -1) {
+        SHERPA_LOG(WARNING) << "Skip OOV token, code point : " << code
+                            << " utf8 char : " << CodePointToUTF8String(code);
+        continue;
+      }
+      bytes->push_back(token2byte_[code]);
+    }
+  }
+}
+
+void ByteUtil::BytesToCodePoints(uint8_t *bytes, int32_t length,
+                                 std::vector<int32_t> *codes) const {
+  if (length <= 0) return;
+  char *data = reinterpret_cast<char *>(bytes);
+  int32_t idx = 1;  // means start from next byte
+  for (int32_t i = 0; i < length; /* no update */) {
+    int32_t c = data[i++] & 0xff;
+    if ((c & 0x80) == 0) {
+      codes->push_back(c);
+      idx = i + 1;
+    } else {
+      if ((c & 0xc0) == 0x80) {
+        BytesToCodePoints(bytes + idx, length - idx, codes);
+        return;
+      }
+      int32_t count =
+          (c >= 0xc0) + (c >= 0xe0) + (c >= 0xf0) + (c >= 0xf8) + (c >= 0xfc);
+      int32_t code = c & ((1 << (6 - count)) - 1);
+      while (count != 0) {
+        if (i == length) {
+          BytesToCodePoints(bytes + idx, length - idx, codes);
+          return;
+        }
+        char cb = data[i++];
+        if ((cb & 0xc0) != 0x80) {
+          BytesToCodePoints(bytes + idx, length - idx, codes);
+          return;
+        }
+        code = (code << 6) | (cb & 0x3f);
+        count--;
+      }
+      if (code < 0) {
+        BytesToCodePoints(bytes + idx, length - idx, codes);
+        return;
+      }
+      codes->push_back(code);
+      idx = i + 1;
+    }
+  }
+}
+
+std::string ByteUtil::CodePointToUTF8String(int32_t code) const {
```

Comment:
Since `code` is in the range [0, 422], can we pre-compute the mapping and save it to `std::vector<std::string>`?
It is more efficient if we pre-compute it.

Reply:
This function is used to convert general unicode points to string too.